### PR TITLE
feat(datasource): ODC connection module supports oracle11g

### LINKED ISSUE #396: [Feature]: ODC support connect to oracle datasource
### Is your feature request related to a problem?

no

### Describe the solution you'd like

ODC support connect to native oracle datasource

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PeachThinking**
#### What type of PR is this?
type-feature
module-Datasource

#### What this PR does / why we need it:
Currently, this branch can create and synchronize Oracle data sources, create sessions, and execute SQL in the SQL window.
1. Connection module support for Oracle11g: add `connect-plugin-oracle`.
2. create new datasource support for Oracle11g:
add `sid`、`serviceName`、`userRole` column in table `connect_connection`, these columns are need for oracle to create a new connection.
3. Sync database and create database support for oracle: add `schema-plugin-oracle`.
4. Create session support oracle: make `DefaultConnectSessionFactory` adapt to oracle. query 
 session variable method in `DefaultConnectSessionFactory#setNlsFormat`  does not work with oracle mode，so a `getVariable` method is abstracted in the `SessionExtensionPoint` interface.
6. query and set `autocommit` variable method in `SessionSettingsService` does not work with oracle mode. because Oracle does not have an `autocommit` session variable, so jdbc is used to obtain and set the `autocommit` of the current connection.    I have verified that oceanbase、mysql and oralce's jdbc all support this method.
```
boolean autoCommit = connection.getAutoCommit();
connection.setAutoCommit(false);
```

7. fix do not connect to correct database when create session by database.  because oracle cannot set defaultSchema in jdbcUrl, so  we cannot get  connect schema by `JdbcUrlParser#getSchema`  in `DefaultConnectSessionFactory#registerConsoleDataSource`. In this case, oracle's `JdbcUrlParser#getSchema` will will return `null`, which means the current connection does not specify a database to connect to, so `ConnectionSessionUtil.setConnectSchema(session, null);`, and `SwitchSchemaInitializer` will switch schema to current schema.

 
#### Which issue(s) this PR fixes:

Fixes #396

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/rollbackplan/RollBackPlanJdbcRowMapper.java [hunk 1]
```diff
@@ -42,13 +42,13 @@ public class RollBackPlanJdbcRowMapper extends BaseDialectBasedRowMapper {
...
         super(dialectType);
-        if (dialectType == DialectType.OB_MYSQL) {
+        if (dialectType.isMysql()) {
             mapperList.add(new MySQLBitMapper());
             mapperList.add(new MySQLDatetimeMapper());
             mapperList.add(new MySQLYearMapper());
             mapperList.add(new MySQLTimestampMapper());
             mapperList.add(new MySQLGeometryMapper());
-        } else if (dialectType == DialectType.OB_ORACLE) {
+        } else if (dialectType.isOracle()) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/rollbackplan/RollBackPlanJdbcRowMapper.java:51]: same as `DefaultJdbcRowMapper`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java [hunk 2]
```diff
@@ -53,16 +53,13 @@ public class SessionSettingsService {
...
-            SessionSettings settings = new SessionSettings();
-            if (rs.next()) {
-                settings.setAutocommit("on".equalsIgnoreCase(rs.getString(2)));
-            }
-            settings.setObVersion(ConnectionSessionUtil.getVersion(session));
-            settings.setDelimiter(ConnectionSessionUtil.getSqlCommentProcessor(session).getDelimiter());
-            settings.setQueryLimit(ConnectionSessionUtil.getQueryLimit(session));
-            return settings;
-        });
+        Boolean autocommit = jdbcOperations.execute(Connection::getAutoCommit);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java:56]: great

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java [hunk 3]
```diff
@@ -72,16 +69,12 @@ public SessionSettings setSessionSettings(@NotNull ConnectionSession session,
...
-        boolean commitFlag = false;
-        if (CollectionUtils.isNotEmpty(lines)) {
-            String autocommit = lines.get(0);
-            commitFlag = "on".equalsIgnoreCase(autocommit);
-        }
-        if (commitFlag != settings.getAutocommit()) {
-            sql = String.format("set session autocommit='%s'", settings.getAutocommit() ? "ON" : "OFF");
-            jdbcOperations.execute(sql);
+        Boolean autocommit = jdbcOperations.execute(Connection::getAutoCommit);
+        if (!autocommit.equals(settings.getAutocommit())) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java:73]: Objects.equal will be better

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java [hunk 4]
```diff
@@ -72,16 +69,12 @@ public SessionSettings setSessionSettings(@NotNull ConnectionSession session,
...
+            jdbcOperations.execute((ConnectionCallback<Void>) conn -> {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java:74]: we may test for oceanbase mode, the test cases can be as follow:
1. set auto commit by front end
2. execute `show variables like '%autocommit%'` to confirm if setting is working

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java [hunk 5]
```diff
@@ -104,10 +104,13 @@ private void registerConsoleDataSource(ConnectionSession session) {
                 connectSchema = ConnectionSessionUtil.getUserOrSchemaString(connectSchema, session.getDialectType());
                 ConnectionSessionUtil.setConnectSchema(session, connectSchema);
                 ConnectionSessionUtil.setCurrentSchema(session, connectSchema);
+            } else {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java:107]: we may change connect extension to solve this problem?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java [hunk 6]
```diff
@@ -104,10 +104,13 @@ private void registerConsoleDataSource(ConnectionSession session) {
...
+                ConnectionSessionUtil.setConnectSchema(session, null);
+                ConnectionSessionUtil.setCurrentSchema(session, connectionConfig.getDefaultSchema());
             }
         } catch (Exception e) {
             if (StringUtils.isNotBlank(connectionConfig.getDefaultSchema())) {
-                ConnectionSessionUtil.setConnectSchema(session, connectionConfig.getDefaultSchema());
+                ConnectionSessionUtil.setConnectSchema(session, null);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java:113]: null value is not a good idea

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DruidDataSourceFactory.java [hunk 7]
```diff
@@ -87,7 +94,7 @@ private void init(DruidDataSource dataSource) {
         dataSource.setSocketTimeout(DEFAULT_TIMEOUT_MILLIS);
         dataSource.setConnectTimeout(DEFAULT_TIMEOUT_MILLIS);
         // fix arbitrary file reading vulnerability
-        Properties properties = new Properties();
+        Properties properties = dataSource.getConnectProperties();
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DruidDataSourceFactory.java:97]: npe?

File: server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/api/ConnectionExtensionPoint.java [hunk 8]
```diff
@@ -33,7 +35,14 @@
...
 
-    String generateJdbcUrl(String host, Integer port, String defaultSchema, Map<String, String> jdbcParameters);
+    /**
+     * @param properties Properties required by jdbcURL, such as HOST, PORT and DEFAULT_SCHEMA, see
+     *        {@link ConnectionConstants}
+     * @param jdbcParameters jdbc parameters.
+     *
+     * @return jdbcURL
+     */
+    String generateJdbcUrl(Properties properties, Map<String, String> jdbcParameters);
```
Comment (yhilmare) [inline server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/api/ConnectionExtensionPoint.java:45]: why should we use `Properties` to input host and port? Is this stipulated in the jdbc protocol?

File: server/plugins/connect-plugin-oracle/pom.xml [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            <groupId>com.oceanbase</groupId>
+            <artifactId>connect-plugin-ob-oracle</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.oceanbase</groupId>
+            <artifactId>connect-plugin-ob-mysql</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
```
Comment (yhilmare) [inline server/plugins/connect-plugin-oracle/pom.xml:56]: add test scope

File: server/plugins/connect-plugin-oracle/pom.xml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.oceanbase</groupId>
+            <artifactId>odc-test</artifactId>
```
Comment (yhilmare) [inline server/plugins/connect-plugin-oracle/pom.xml:60]: same as above

File: server/plugins/connect-plugin-oracle/src/main/java/com/oceanbase/odc/plugin/connect/oracle/OracleConnectionExtension.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Override
+    public List<ConnectionInitializer> getConnectionInitializers() {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public TestResult test(@NonNull String jdbcUrl, @NonNull Properties properties, int queryTimeout) {
+        try (Connection connection = DriverManager.getConnection(jdbcUrl, properties)) {
```
Comment (yhilmare) [inline server/plugins/connect-plugin-oracle/src/main/java/com/oceanbase/odc/plugin/connect/oracle/OracleConnectionExtension.java:97]: just a suggestion: use JdbcOperations

File: server/plugins/task-plugin-mysql/src/main/java/com/oceanbase/odc/plugin/task/mysql/datatransfer/MySQLDataTransferJob.java [hunk 12]
```diff
@@ -172,15 +174,29 @@ private HikariDataSource initDataSource() {
...
+        dataSource.setJdbcUrl(
+                new MySQLConnectionExtension().generateJdbcUrl(getJdbcUrlProperties(connectionInfo), jdbcUrlParams));
         dataSource.setUsername(connectionInfo.getUserNameForConnect());
         dataSource.setPassword(connectionInfo.getPassword());
         dataSource.setDriverClassName(OdcConstants.MYSQL_DRIVER_CLASS_NAME);
         dataSource.setMaximumPoolSize(3);
         return dataSource;
     }
 
+    private Properties getJdbcUrlProperties(ConnectionInfo connectionInfo) {
```
Comment (yhilmare) [inline server/plugins/task-plugin-mysql/src/main/java/com/oceanbase/odc/plugin/task/mysql/datatransfer/MySQLDataTransferJob.java:186]: dup codes

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java (hunk 4 above)
Comment (PeachThinking) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/SessionSettingsService.java:74]: Passed the test following this step

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java (hunk 5 above)
Comment (PeachThinking) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java]: solve it by add userName parameter in `JdbcUrlParser getConnectionInfo(@NonNull String jdbcUrl, String userName) method`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java (hunk 6 above)
Comment (PeachThinking) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java]: modified.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java [hunk 13]
```diff
@@ -98,13 +98,13 @@ private void registerConsoleDataSource(ConnectionSession session) {
         try {
             JdbcUrlParser urlParser = ConnectionPluginUtil
                     .getConnectionExtension(connectionConfig.getDialectType())
-                    .getJdbcUrlParser(dataSourceFactory.getJdbcUrl());
+                    .getConnectionInfo(dataSourceFactory.getJdbcUrl(), connectionConfig.getUsername());
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java:101]: dataSourceFactory.getUsername() will be better

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java [hunk 14]
```diff
@@ -98,13 +98,13 @@ private void registerConsoleDataSource(ConnectionSession session) {
...
             String connectSchema = urlParser.getSchema();
             if (StringUtils.isNotBlank(connectSchema)) {
                 connectSchema = ConnectionSessionUtil.getUserOrSchemaString(connectSchema, session.getDialectType());
                 ConnectionSessionUtil.setConnectSchema(session, connectSchema);
-                ConnectionSessionUtil.setCurrentSchema(session, connectSchema);
             }
+            ConnectionSessionUtil.setCurrentSchema(session, connectionConfig.getDefaultSchema());
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DefaultConnectSessionFactory.java:107]: what if current schema different from connectSchema?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DruidDataSourceFactory.java [hunk 15]
```diff
@@ -60,6 +63,11 @@ public DataSource getDataSource() {
         dataSource.setUrl(jdbcUrl);
         dataSource.setUsername(username);
         dataSource.setPassword(password);
+        if (Objects.nonNull(this.userRole)) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/DruidDataSourceFactory.java:66]: use builder will be better?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/OBConsoleDataSourceFactory.java [hunk 16]
```diff
@@ -176,10 +194,15 @@ public DataSource getDataSource() {
         dataSource.setUrl(jdbcUrl);
         dataSource.setUsername(username);
         dataSource.setPassword(password);
+
+        Properties properties = new Properties();
+        if (Objects.nonNull(this.userRole)) {
+            properties.put(ConnectionPropertiesBuilder.USER_ROLE, this.userRole.name());
+            dataSource.setConnectionProperties(properties);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/factory/OBConsoleDataSourceFactory.java:201]: `dataSource.setConnectionProperties(properties);` is not necessary

File: server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/model/UserRole.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+ * limitations under the License.
+ */
+package com.oceanbase.odc.plugin.connect.model;
+
+/**
+ * @author jingtian
+ * @date 2023/11/13
+ * @since
+ */
+public enum UserRole {
```
Comment (yhilmare) [inline server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/model/UserRole.java:23]: this should only for oracle, may: `server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/model/oracle/UserRole.java`

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**